setup.sh

```diff
@@ -0,0 +1,395 @@
+	ip -o link show | awk -F': ' '{print $2}' | sed 's/[ \t].*//;/^\(lo\|\)$/d' | sed 's/[ \t].*//;/^\(bond0\|\)$/d' | grep "$tink_interface"
+	
+	if [ $? -ne 0 ]; then
+		echo "$ERR Invalid interface selected. Exiting setup."
+		exit 1
+	fi	
+	echo -e "# Network interface for Tinkerbell \nexport TINKERBELL_NETWORK_INTERFACE=$tink_interface" >> "$ENV_FILE"	
+}
+
+get_tinkerbell_ips() {
+	read -p 'Select the subnet for Tinkerbell ecosystem: [default 192.168.1.0/29]: ' tink_network
+	tink_network=${tink_network:-"192.168.1.0/29"}
+
+	read -p 'Select the IP address for Tinkerbell [default 192.168.1.1]: ' ip_addr	
+	ip_addr=${ip_addr:-"192.168.1.1"}	
+
+	host=$(($(echo $ip_addr | cut -d "." -f 4 | xargs) + 1))
+	nginx_ip="$(echo $ip_addr | cut -d "." -f 1).$(echo $ip_addr | cut -d "." -f 2).$(echo $ip_addr | cut -d "." -f 3).$host"
+
+	# calculate network and broadcast based on supplied provide IP range
+	if [[ $tink_network =~ ^([0-9\.]+)/([0-9]+)$ ]]; then
+		# CIDR notation
+		IPADDR=${BASH_REMATCH[1]}
+		NETMASKLEN=${BASH_REMATCH[2]}
+		zeros=$((32-NETMASKLEN))
+		NETMASKNUM=0
+		for (( i=0; i<$zeros; i++ )); do
+			NETMASKNUM=$(( (NETMASKNUM << 1) ^ 1 ))
+		done
+		NETMASKNUM=$((NETMASKNUM ^ 0xFFFFFFFF))
+		toaddr $NETMASKNUM NETMASK
+	else
+    	IPADDR=${1:-192.168.1.1}
+    	NETMASK=${2:-255.255.255.248}
+	fi
+
+	tonum $IPADDR IPADDRNUM
+	tonum $NETMASK NETMASKNUM
+
+	# The logic to calculate network and broadcast
+	INVNETMASKNUM=$(( 0xFFFFFFFF ^ NETMASKNUM ))
+	NETWORKNUM=$(( IPADDRNUM & NETMASKNUM ))
+	BROADCASTNUM=$(( INVNETMASKNUM | NETWORKNUM ))
+
+	toaddr $NETWORKNUM NETWORK
+	toaddr $BROADCASTNUM BROADCAST
+
+	echo -e "\n# Subnet (IP block) used by Tinkerbell ecosystem \nexport TINKERBELL_NETWORK=$tink_network" >> "$ENV_FILE"	
+	echo -e "\n# Host IP is used by provisioner to expose different services such as tink, boots, etc. \nexport TINKERBELL_HOST_IP=$ip_addr" >> "$ENV_FILE"
+	echo -e "\n# NGINX IP is used by provisioner to serve files required for iPXE boot \nexport TINKERBELL_NGINX_IP=$nginx_ip" >> "$ENV_FILE"	
+	echo -e "\n# Netmask for Tinkerbell network \nexport TINKERBELL_NETMASK=$NETMASK" >> "$ENV_FILE"	
+	echo -e "\n# Broadcast IP for Tinkerbell network \nexport TINKERBELL_BROADCAST_IP=$BROADCAST" >> "$ENV_FILE"	
+}
+
+tonum() {
+    if [[ $1 =~ ([[:digit:]]+)\.([[:digit:]]+)\.([[:digit:]]+)\.([[:digit:]]+) ]]; then
+        addr=$(( (${BASH_REMATCH[1]} << 24) + (${BASH_REMATCH[2]} << 16) + (${BASH_REMATCH[3]} << 8) + ${BASH_REMATCH[4]} ))
+        eval "$2=\$addr"
+    fi
+}
+
+toaddr() {
+    b1=$(( ($1 & 0xFF000000) >> 24))
+    b2=$(( ($1 & 0xFF0000) >> 16))
+    b3=$(( ($1 & 0xFF00) >> 8))
+    b4=$(( $1 & 0xFF ))
+    eval "$2=\$b1.\$b2.\$b3.\$b4"
+}
+
+get_registry_credentials() {
+	read -p 'Create a Docker registry username [default admin]? ' username	
+	username=${username:-"admin"}
+
+	read -sp 'Registry password [default admin]? ' password	
```

Comment:
You may want to generate this instead, see how we do this in openfaas -> https://github.com/openfaas/faas-netes/tree/master/chart/openfaas#generate-basic-auth-credentials

The `read` statements ideally should read from env-vars or a similar mechanism that doesn't require user interaction.

piping or `cat`-ing a file is probably not going to be a robust option because it depends on the ordering of the file and is hard to document. It is also sensitive to timing bugs.

Reply:
Example:

```
if [ $TB_PASSWORD == "" ];
  read -sp 'Registry password [default admin]? ' TB_PASSWORD
fi
echo The password is $TB_PASSWORD
```

Then use it via:

```
# Interactive
curl https:// | bash
# inline value
curl https:// | TB_PASSWORD="test1234"  bash
# or
# Set on separate lines
export TB_PASSWORD="test1234"
curl https:// |  bash
```